### PR TITLE
[SILOptimizer][DebugInfo] Preliminary support for DIExpression in SROA and Mem2Reg

### PR DESCRIPTION
**Pull request by mshockwave**
SROA and Mem2Reg now can leverage DIExpression -- op_fragment, more
specifically -- to generate correct debug info for optimized SIL. Some
important highlights:
  - The new swift::salvageDebugInfo, similar to llvm::salvageDebugInfo,
     tries to restore / transfer debug info from a deleted instruction.
     Currently I only implemented this for store instruction whose
     destination is an alloc_stack value.
 - DCE now does not remove `debug_value` instruction whose associating
     with a function argument SSA value that is not used elsewhere. Since
     that SSA value will not disappear so we should keep the debug info.

**NOTE**: This PR is only for code review. It depends on and _includes_ #38734  #38735 . Please merge this PR instead of the previous two.

### CODE REVIEW

**Comment by mshockwave** [comment]
@swift-ci please test

File: include/swift/SIL/SILCloner.h [hunk 1]
```diff
@@ -748,9 +748,13 @@ SILCloner<ImplClass>::visitAllocStackInst(AllocStackInst *Inst) {
     Loc = MandatoryInlinedLocation::getAutoGeneratedLocation();
     VarInfo = None;
   }
-  recordClonedInstruction(Inst, getBuilder().createAllocStack(
-                                    Loc, getOpType(Inst->getElementType()),
-                                    VarInfo, Inst->hasDynamicLifetime()));
+  auto *NewInst =
+      getBuilder().createAllocStack(Loc, getOpType(Inst->getElementType()),
+                                    VarInfo, Inst->hasDynamicLifetime());
+  // Remap the auxiliary debug scope, if any
```
Comment (adrian-prantl) [inline include/swift/SIL/SILCloner.h:754]: Nit: missing `.` at the end of sentence.

File: include/swift/SIL/SILCloner.h [hunk 2]
```diff
@@ -748,9 +748,13 @@ SILCloner<ImplClass>::visitAllocStackInst(AllocStackInst *Inst) {
...
+  if (VarInfo && VarInfo->Scope)
+    NewInst->setDebugVarScope(getOpScope(VarInfo->Scope));
```
Comment (adrian-prantl) [inline include/swift/SIL/SILCloner.h:756]: Should this be factored into a helper function that does this for all instructions that carry a SILDebugVariable?
I.e., `remapVarInfo(NewInst, VarInfo);`

File: include/swift/SILOptimizer/Utils/DebugOptUtils.h [hunk 3]
```diff
@@ -76,6 +81,7 @@ eraseFromParentWithDebugInsts(SILInstruction *inst,
   // Just matching what eraseFromParentWithDebugInsts is today.
   if (nextII == inst->getIterator())
     ++nextII;
+  swift::salvageDebugInfo(inst);
```
Comment (adrian-prantl) [inline include/swift/SILOptimizer/Utils/DebugOptUtils.h:84]: Does this need to be part of the SROA PR or should this be in its own patch?

File: lib/IRGen/IRGenDebugInfo.cpp [hunk 4]
```diff
@@ -191,9 +191,11 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
   void emitArtificialFunction(IRBuilder &Builder, llvm::Function *Fn,
                               SILType SILTy);
 
-  void handleFragmentDIExpr(const SILDIExprOperand &CurDIExprOp,
+  /// Return false if we fail to create the right DW_OP_LLVM_fragment operand
```
Comment (adrian-prantl) [inline lib/IRGen/IRGenDebugInfo.cpp:194]: `.`

File: lib/IRGen/IRGenDebugInfo.cpp [hunk 5]
```diff
@@ -191,9 +191,11 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
...
+  bool handleFragmentDIExpr(const SILDIExprOperand &CurDIExprOp,
                             SmallVectorImpl<uint64_t> &Operands);
-  void buildDebugInfoExpression(const SILDebugVariable &VarInfo,
+  /// Return false if we fail to create the desired !DIExpression
```
Comment (adrian-prantl) [inline lib/IRGen/IRGenDebugInfo.cpp:197]: `.`

File: lib/IRGen/IRGenDebugInfo.cpp [hunk 6]
```diff
@@ -191,9 +191,11 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
   void emitArtificialFunction(IRBuilder &Builder, llvm::Function *Fn,
                               SILType SILTy);
 
-  void handleFragmentDIExpr(const SILDIExprOperand &CurDIExprOp,
+  /// Return false if we fail to create the right DW_OP_LLVM_fragment operand
+  bool handleFragmentDIExpr(const SILDIExprOperand &CurDIExprOp,
                             SmallVectorImpl<uint64_t> &Operands);
```
Comment (adrian-prantl) [inline lib/IRGen/IRGenDebugInfo.cpp:196]: ArrayRef<uint64_t> ?

File: lib/IRGen/IRGenDebugInfo.cpp [hunk 7]
```diff
@@ -2355,7 +2356,9 @@ void IRGenDebugInfoImpl::handleFragmentDIExpr(
       emitPhysicalStructMemberFixedOffset(IGM, ParentSILType, VD);
   llvm::Type *FieldTy =
       getPhysicalStructFieldTypeInfo(IGM, ParentSILType, VD)->getStorageType();
-  assert(Offset && FieldTy && "Non-fixed type?");
+  // Doesn't support non-fixed type right now
+  if (!Offset || !FieldTy)
+    return false;
```
Comment (adrian-prantl) [inline lib/IRGen/IRGenDebugInfo.cpp:2361]: I'd be curious to see an example of the IR here. Even for a non-fixed size type, I would expect us to emit some sort of fixed-size container that optionally contains a pointer to a heap/stack-allocated memory, and I would expect SROA to only split apart that fixed-size container and not the variable-sized part. We need to fix this right now, but maybe keep this in mind.

File: lib/IRGen/IRGenDebugInfo.cpp [hunk 8]
```diff
@@ -2366,22 +2369,26 @@ void IRGenDebugInfoImpl::handleFragmentDIExpr(
...
   const auto &DIExpr = VarInfo.DIExpr;
   for (const SILDIExprOperand &ExprOperand : DIExpr.operands()) {
     switch (ExprOperand.getOperator()) {
     case SILDIExprOperator::Fragment:
-      handleFragmentDIExpr(ExprOperand, Operands);
+      if (!handleFragmentDIExpr(ExprOperand, Operands))
+        return false;
       break;
     default:
       llvm_unreachable("Unrecoginized operator");
```
Comment (adrian-prantl) [inline lib/IRGen/IRGenDebugInfo.cpp:2388]: not your code: typo

File: lib/IRGen/IRGenSIL.cpp [hunk 9]
```diff
@@ -4752,8 +4752,7 @@ void IRGenSILFunction::visitDebugValueInst(DebugValueInst *i) {
   if (VarDecl *Decl = i->getDecl()) {
     DbgTy = DebugTypeInfo::getLocalVariable(
         Decl, RealTy, getTypeInfo(SILVal->getType()));
-  } else if (i->getFunction()->isBare() && !SILTy.hasArchetype() &&
-             !VarInfo->Name.empty()) {
+  } else if (!SILTy.hasArchetype() && !VarInfo->Name.empty()) {
```
Comment (adrian-prantl) [inline lib/IRGen/IRGenSIL.cpp:4755]: why the change with isBare?

File: lib/SIL/IR/SILBasicBlock.cpp [hunk 10]
```diff
@@ -292,8 +292,17 @@ transferNodesFromList(llvm::ilist_traits<SILBasicBlock> &SrcTraits,
...
-    for (auto &II : *First)
+    for (auto &II : *First) {
       II.setDebugScope(ScopeCloner.getOrCreateClonedScope(II.getDebugScope()));
+      // Special handling for SILDebugVariable
+      if (auto DVI = DebugVarCarryingInst(&II)) {
+        if (auto VarInfo = DVI.getVarInfo()) {
+          if (VarInfo->Scope)
+            DVI.setDebugVarScope(
+                ScopeCloner.getOrCreateClonedScope(VarInfo->Scope));
+        }
```
Comment (adrian-prantl) [inline lib/SIL/IR/SILBasicBlock.cpp:303]: LLVM style usually omits {} for single statements.

File: lib/IRGen/IRGenSIL.cpp (hunk 9 above)
Comment (mshockwave) [reply lib/IRGen/IRGenSIL.cpp:4755]: A function is marked as bare if it's parsed from SIL. But I also saw cases where the function is coming from Swift code but the instruction (i.e. `DebugValueInst *I`) doesn't have an associated VarDecl. For example, the debug_value instruction was generated during the salvageDebugInfo. In those cases, we will restore the `DbgTy` from `RealTy` and `SILTy`.

File: include/swift/SILOptimizer/Utils/DebugOptUtils.h (hunk 3 above)
Comment (mshockwave) [reply include/swift/SILOptimizer/Utils/DebugOptUtils.h:84]: My original thought was that if salvageDebugInfo is put in a separate patch (or PR), there will be no use case or test there, which is a little weird. What do you think?

Comment (adrian-prantl) [reply include/swift/SILOptimizer/Utils/DebugOptUtils.h:84]: Could we write a standalone test using sil-opt that runs exactly one pass that would delete an instruction? If the answer is no, then let's just keep it.

File: lib/IRGen/IRGenDebugInfo.cpp (hunk 6 above)
Comment (adrian-prantl) [reply lib/IRGen/IRGenDebugInfo.cpp:196]: Sorry, I didn't realize that.

File: lib/IRGen/IRGenSIL.cpp (hunk 9 above)
Comment (adrian-prantl) [reply lib/IRGen/IRGenSIL.cpp:4755]: Under what circumstances does salvageDebugInfo drop the VarDecl? Shouldn't it carry it forward from the salvaged debug_value?

Comment (mshockwave) [reply lib/IRGen/IRGenSIL.cpp:4755]: We're actually retrieving a VarDecl of a debug_value from its attached SILLocation (`getLoc().getAsASTNode<VarDecl>();`). And in the case of debug_value created by salvageDebugInfo, its SILLocation is "<compiler generated>", which I believe we can not get it as VarDecl.

File: include/swift/SILOptimizer/Utils/DebugOptUtils.h (hunk 3 above)
Comment (mshockwave) [reply include/swift/SILOptimizer/Utils/DebugOptUtils.h:84]: Currently we're calling salvageDebugInfo inside InstructionDeleter, which will affect multiple Passes. We _can_ first create a PR that only calls salvageDebugInfo in a single Pass then revise it to use InstructionDeleter in later PR...but I feel like it will be a little over-engineering

**Comment by mshockwave** [comment]
I just found that I might can split the changes regarding debug scope here to a separate PR. I'll update this accordingly later

File: lib/IRGen/IRGenSIL.cpp (hunk 9 above)
Comment (adrian-prantl) [reply lib/IRGen/IRGenSIL.cpp:4755]: Thanks! We discussed this offline and the piece of info I was missing that we are storing the type info in the new auxiliary info record, and this change effectively allows us to access that.

**Comment by mshockwave** [comment]
> I just found that I might can split the changes regarding debug scope here to a separate PR. I'll update this accordingly later

Since that part of changes don't really affect the feature here, I'll put it in another PR that is not depending on the current series of patches
@swift-ci please test

**Comment by mshockwave** [comment]
@swift-ci please test

**Comment by mshockwave** [comment]
> > I just found that I might can split the changes regarding debug scope here to a separate PR. I'll update this accordingly later
> 
> Since that part of changes don't really affect the feature here, I'll put it in another PR that is not depending on the current series of patches

I was wrong about that: Without the debug scope changes it will fail on a pretty common optimization pattern (specialization on implicit-generated enum functions) despite passing all tests, so I'm bringing them back and add corresponding test cases. I've also addressed all the feedbacks on that part before they were previously detached from this PR.
Sorry for this.

@swift-ci please test

**Review by adrian-prantl — APPROVED** [review]
Great, I think you've addressed all my concerns.

**Comment by mshockwave** [comment]
So the failure on Linux buildbot was caused by uninitialized struct type information during IRGen. I think it only happens if a debug_value instruction is the first ever SIL instruction that references a certain struct type, in which case IRGen hasn't translated that type into its internal representation (e.g. RecordTypeInfo).

A more sophisticated solution will be asking IRGen to resolve the type when this happens. But since this scenario is pretty rare and I don't want to add more complexity into this PR, for now I simply bails out the (LLVM) debug intrinsic generation process if this occurs, similar to our current resolution on non-fixed type.

@swift-ci please test

**Comment by adrian-prantl** [comment]
> A more sophisticated solution will be asking IRGen to resolve the type when this happens. But since this scenario is pretty rare and I don't want to add more complexity into this PR, for now I simply bails out the (LLVM) debug intrinsic generation process if this occurs, similar to our current resolution on non-fixed type.

Thanks for adding the comment explaining this to IRGenDebugInfo. LGTM!
